### PR TITLE
[Quantization] Add histogram to profile and KL calibration method.

### PR DESCRIPTION
**Pull request by mciprian13**
**Summary**
- Add the histogram to the profile (besides current min/max).
- Add quantization range optimization method based on minimizing the KL divergence metric. The method is inspired from MXNet/TVM but the current code changed quite a bit from the original because it was cleaned, commented, fixed (the original had an issue) and optimized 2-3x.
- Added extra options to loader: calibration mode (none/KL), flag to calibrate also the constant weighths, choose the number of histogram bins.
- Extra Python utility to plot a histogram for a node-value using the profile file.

**Documentation**
At this point I want to add this feature for experimentation. I`m not sure what is the sweet spot in terms of the algorithm parameters:
- what is the optimal number of histogram bins to use (TVM uses 8001 but I think that is too much because the calibration is O(N^2))
- whether it makes sense to calibrate also the constant weights
Following some experimentation in the following days/weeks I might add a more comprehensive documentation about quantization/calibration tuning.

**Test Plan**
Manually tested the method against the original implementation from MXNet/TVM.
Added some simple unit tests to check some corner cases.
I will test this feature with a couple of models in the next period to see how it behaves.
I ran the calibration on the MobileNet v1 1.0 224 just to check two things:
- the profile file size: ~1.2 MB which is okay (including the histogram with 1000 bins)
- the compile time for the model when quantizing it: a couple of seconds which is okay
I will try in the following period to test whether this method helps some models to have better accuracy when quantized.

### CODE REVIEW

File: lib/Quantization/Base/Base.cpp [hunk 1]
```diff
@@ -389,6 +391,31 @@ chooseQuantizationParams(TensorProfilingParams profParams, Schema schema,
   min = std::max(min, std::numeric_limits<float>::lowest());
   max = std::min(max, std::numeric_limits<float>::max());
 
+  // Calibrate the min/max range (for non-zero ranges only).
+  if ((profParams.min != profParams.max) && (min != max)) {
+    if (calibration == Calibration::KLMinimization) {
```
Comment (jfix71) [inline lib/Quantization/Base/Base.cpp:396]: nit: combine into a single if so there's one less level of nesting?

File: include/glow/Quantization/Base/Base.h [hunk 2]
```diff
@@ -79,6 +94,9 @@ struct NodeProfilingInfo {
 
   float Min() const { return tensorProfilingParams_.min; }
   float Max() const { return tensorProfilingParams_.max; }
+  const std::vector<float> &Histogram() const {
+    return tensorProfilingParams_.histogram;
+  }
```
[the remark this answers is not in the document]
Comment (mciprian13) [reply include/glow/Quantization/Base/Base.h]: Done. I did also for the "Scale()" and "Offset()" of the NodeQuantizationInfo structure.

File: lib/Quantization/Quantization.cpp [hunk 3]
```diff
@@ -708,23 +708,20 @@ class FunctionQuantizer : public FunctionConverter {
   /// converted. \p assertAllNodesQuantized is used as a debugging tool; if
   /// true then if the backend does not support a node as quantized for the
   /// given \p quantizationPrecision then the program will exit with an error.
-  FunctionQuantizer(Function &F, const Backend &B, quantization::Schema schema,
-                    llvm::ArrayRef<NodeProfilingInfo> profilingInfos,
-                    ElemKind quantizationPrecision,
+  FunctionQuantizer(Function &F, const Backend &B,
```
[the remark this answers is not in the document]
Comment (mciprian13) [reply lib/Quantization/Quantization.cpp:710]: Done.

File: lib/Quantization/Base/Base.cpp (hunk 1 above)
Comment (mciprian13) [reply lib/Quantization/Base/Base.cpp]: Done.

File: lib/Quantization/Quantization.cpp [hunk 4]
```diff
@@ -1005,16 +1007,29 @@ generateNodeQuantizationInfos(llvm::ArrayRef<NodeProfilingInfo> profilingInfos,
...
     }
 
+    // Do not calibrate the quantization parameters for scalars.
+    if (nodeOutput.getType()->size() == 1) {
+      calibration = Calibration::None;
+    }
+
+    // Disable the quantization calibration for constant weights.
+    if (!quantConfig.calibrateConstants &&
+        llvm::dyn_cast<Constant>(nodeOutput.getNode())) {
```
[the remark this answers is not in the document]
Comment (mciprian13) [reply lib/Quantization/Quantization.cpp]: Done.

**Comment by mciprian13** [comment]
@jfix I ran the profiling and got the following profile file sizes:
- Resnet50v2 -> 5.45 MB
- MobileNetv2 -> 1.8 MB
I still think this is ok since the size of the profile is correlated with the number of nodes and not exactly the model size. If this seems too much a simple solution would be to change the default number of bins from 1000 to (say) 10 in Loader. What do you think?

**Comment by mciprian13** [comment]
@jfix71 Side question: can you have your say in the PR #4350 which is becoming stale?
I refactored the ChannelwiseQuantizedConvolution but right before I created the PR another PR merged the definitions ChannelwiseQuantizedConvolution with the ChannelwiseQuantizedConv3D. There are some on-going discussions on whether this is the right thing to do and whether we should also merge the definitions of float Conv and Conv3D for consistency ...

**Comment by jfix71** [comment]
>If this seems too much a simple solution would be to change the default number of bins from 1000 to (say) 10 in Loader. What do you think?

Yes I think this is a reasonable solution, or even set it to 0 by default. Just would rather not make the size of the profile super large unnecessarily. Thanks!

**Comment by mciprian13** [comment]
When using 0 histogram bins the compile-time crashes because the `profileQuantization` function creates a tensor with size 0 for the histogram which seems it is not allowed (the assert from Tensor.h:475 crashes). How do you want to proceed with this?

**Comment by mciprian13** [comment]
@jfix71 I added in the histogram related functions a graceful return when the histogram is empty (0 bins) and I changed the default histogram size to 10. So now the only blocking thing of having empty histograms is the restriction to have Tensors with 0 size).

**Comment by jfix71** [comment]
@mciprian13 I wonder if a better solution here is to keep the number of bins high by default, but then also by default skip adding the histogram values to the proto? I don't think it's super clean to skip adding it to the yaml as is, but we could instead add something like the following to `serializeProfilingInfosToYaml()`:

```c++
if (!includeHistogramInYaml) {
  for (auto &i : info) {
    i.tensorProfilingParams_.histogram.clear();
  }
}
```
WDYT?

**Comment by mciprian13** [comment]
@jfix71 I'm not sure I would like having additional flags / knobs just to turn off the histogram because it then makes error-prone turning it on (for somebody it would be frustrating to test the calibration for 10,100,1000 bins with nothing happening and then discovering that he did not used the flag to actually export the histogram).

I still think it does not harm to have the histogram with 10 bins because:
- before this PR the histogram was always computed/gathered during profiling with a default value of 1000 bins and nobody complained about the run time or memory. Having 10 bins would reduce both.
- having 10 bins histogram instead of 0 bins actually does not increase the profile file size that much: only around 30%.
- a histogram with 10 bins is actually more "visually" (for the human eye) suggestive than a histogram of 1000 bins about the data distribution 

If not, I still think the easiest way to go would be to allow zero-size Tensors (in the same way all the containers allow for empty content). What would be the downside of this? Why enforce non-zero size Tensors?
I actually dealt with this problem multiple times: for example when a scratch operand is unconditionally defined for an operator but the scratch is not always required (e.g. scratch might be required for the operation only when the input is higher than 10 elements).
How would you treat this case? At that time I did an ugly hack like: if no scratch is required, allocate 1 dummy byte such that Tensor.h does not complain. Definitely this is not elegant ...

**Comment by jfix71** [comment]
@mciprian13 I have mixed feelings about Tensors of size zero. The way things are currently designed in our Tensors and Types, it'd be nontrivial to allow it. E.g. we have zero-dimensional Tensors as a concept to represent scalars, but they have size 1 as you'd expect, so the combination of these two ideas is going to get messy. If you're interested in this then feel free to open an issue to discuss further, I think it's outside of the scope of this PR.

Let's just move forward with 10 as the default. Can you set that to be the case in `ProfilingConfiguration`? We don't always run profiling through the Loader so right now those uses would also have very large profiles.

**Comment by mciprian13** [comment]
@jfix71 I changed the number of histogram bins to 10 everywhere (Loader, ProfilingConfig and also in the Graph.h when creating the profiling node). Let`s move on with this and have this landed.
Thanks!

**Comment by mciprian13** [comment]
@jfix71 Can we have this landed? Thanks!
